README.rst
########
Overview
########
In order to manage a network, an administrator must have updated and reliable
statistics about it, in several levels of deepness, from a
switch port inbound and outbound traffic to the traffic of different connected
networks.

To achieve that, this Network Application collects statistical data provided by
the switches connected to the controller. We *do not use SNMP protocol* because
the OpenFlow protocol already provide this data. The data is stored to be
provided later through a REST API. This API can supply instant data,
historical data, and also some calculated information.

The provided statistics, per switch, are:

* **Ports/Interfaces**: bytes/sec, utilization, dropped packets/sec and
  errors/sec split into transmission and reception; interface name, MAC address
  and link speed (Bps);
* **Flows**: packets/sec, bytes/sec;

##########
Installing
##########

************
RRDTool
************
.. note:: We currently use rrd to keep persistence in data, but future
    versions of this napp will allow you to choose what kind of backend you
    want to use.

==============
Linux packages
==============
Currently, this NApp uses `rrdtool <http://www.rrdtool.org>`__ and its Python
wrapper needs several packages to be installed and run. In Ubuntu, you can
install them by:

.. code-block:: shell

   apt install rrdtool python3.6-dev librrd-dev

===============
Python packages
===============
With the Linux packages above installed, just run:

.. code-block:: shell

   pip install rrdtool

*****************
NApp installation
*****************
All of the Kytos Network Applications are located in the online NApps
repository. To install this NApp, run:

.. code-block:: shell

   kytos napps install kytos/of_stats


.. code:: shell

   kytos napps install kytos/topology

**********
Developers
**********
To run the NApp from the source code, install the requirements and run:

.. code-block:: shell

   git clone https://github.com/kytos/of_stats.git
   cd of_stats
   kytos napps install kytos/of_stats

With the setup above, ``git pull`` will update NApp.

###########
Configuring
###########
There's no need to configure anything for this NApp to run. Read the sections
below if you need to change of_stats behavior or just want to know more details
on how it works.

*************
Settings file
*************
You can customize settings like how long to wait before asking the switches
for more statistics, as well as rrdtool-related configuration in the file
``settings.py``.

****************
Custom bandwidth
****************
Sometimes, the link speed is wrongly reported by the switch or there's no such
speed in the protocol specification. In these cases, you can manually define
the speeds in the file ``user_speed.json``. Changes to this file will be loaded
automatically without the need to restart the controller.

Setting interface speeds manually is quite easy. Just create
``user_speed.json`` by copying and customizing the provided
``user_speed.example.json`` file. Let's see what the provided example means:

.. code-block:: json

   {
     "default": 12500000000,
     "00:00:00:00:00:00:00:01":
     {
       "default": 1250000000,
       "4": 125000000
     }
   }

Keep in mind that, for consistency reasons, we try to keep the units in bytes
whenever possible. Thus, speeds should specified in bytes/sec (not necessarily
integers).

Any value in this file overrides the OpenFlow values returned by the switches.
In this file, inner values take precedence over outer ones.

The *default* values are optional. The first line has a default value that
specifies the speed of any interface that is not found in this file
(12,500,000,000 Bps = 100 Gbps). Then, there's a default value set for all
interfaces of the switch whose dpid is *00:...:00:01* (1,250,000,000 Bps = 10
Gbps). Even more specifically, its interface with port number 4 is 125,000,000
Bps (1 Gbps).

To make it even more clear, find below the speed of several interfaces when
``user_speed.json`` has the content above:

+-------------------------+------+--------------+
|          DPID 1         | Port | Speed (Gbps) |
+=========================+======+==============+
| 00:00:00:00:00:00:00:01 |  4   |        1     |
+-------------------------+------+--------------+
| 00:00:00:00:00:00:00:01 |  2   |       10     |
+-------------------------+------+--------------+
| 00:00:00:00:00:00:00:02 |  4   |      100     |
+-------------------------+------+--------------+
| 00:00:00:00:00:00:00:02 |  2   |      100     |
+-------------------------+------+--------------+

######
Events
######

********
Listened
********

================================================
kytos/of_core.v0x01.messages.in.ofpt_stats_reply
================================================
This event contains the statistics to be processed.

Content
-------
A KytosEvent object containing:

- message: a `StatsReply` object;
- source: contains the switch datapath ID in ``source.switch.dpid``.

########
Rest API
########
You can find a list of the available endpoints and example input/output in the
'REST API' tab in this NApp's webpage in the `Kytos NApps Server
<https://napps.kytos.io/kytos/of_stats>`_.

###############
Troubleshooting
###############
.. attention:: The filenames below are relative to this NApp's folder.
   If you run Kytos as root, it is ``/var/lib/kytos/napps/kytos/of_stats`` or,
   if using virtualenv, ``$VIRTUAL_ENV/var/lib/kytos/napps/kytos/of_stats``.

**********************
Wrong link utilization
**********************
Check whether the link bandwidth is correct. If it is not, set the correct
bandwidth by following the instructions in *Configuring*, *Custom bandwidth*.

****************************
New settings are not applied
****************************
Some changes in ``settings.py`` require recreating the database. Check the
section ``Deleting the database`` below.

******************
Unexpected results
******************
If you get fewer points than expected for a specific flow or port, it might be
that there's not enough data to calculate all of them. Try to specify the
"start" parameter using a later time that you know traffic was being collected.

You may also try to delete RRD databases (see below) so they can be recreated
with the latest configuration.

*********************
Deleting the database
*********************
You don't have to stop the controller to delete the databases. This NApp will
recreate them as needed after you run (no need to restart the NApp):

.. code-block:: shell

   rm -rf rrd/flows rrd/ports



CHANGELOG.rst
#########
Changelog
#########
All notable changes to the kytos-napps project will be documented in this file.

[UNRELEASED] - Under development
********************************
Added
=====

Changed
=======

Deprecated
==========

Removed
=======

Fixed
=====

Security
========

[1.1.0] - 2018-04-20
********************
Added
=====
- Add speed to port statistics..
- Multipart requests for switches with OpenFlow 1.3.
- Use flow factory.
- Add support for OF 0x04 (1.3).

Changed
=======
- Adapt the NApp to changes in python-openflow.
- Get stats updates from OF1.3 switches.
- Update user_speed to bytes.
- Move tests to tests folder.
- Update controller's custom interface speed.
- Update OpenAPI.
- Standardize interface speed value.
- Return 30 points instead of 60.
- Get latest 30 points instead of whole BD period.
- Update v0x04 body_type to multipart_type.
- Update RRD creation time steps.
- Use generic flow.
- Add /random for random port statistics.
- Update controller's interface stats.
- Update controller's flow.stats.

Fixed
=====
- Fix retrieve flows error.
- Fix exception when disabling NApp.
- Fix RRD-not-found reply.
- Fix bug when switch has no speed.

[1.0.0] - 2017-11-30
********************
Added
=====
- Use new generic Flow from of_core.
- Create Rest API in openapi.yml.
- Use rest decorator in of_stats.
- Adding dependency of `kytos/of_core` in kytos.json.
- include __init__ files in version specific of_core utils modules.

[0.4.0] - 2017-06-16
********************
Added
=====
- Experimental OpenFlow 1.3 support in of_core.
- Examples of requests/replies to of_flow_manager REST endpoints.

Changed
=======
- of_core is now responsible for ALL basic OpenFlow packet handling, previously
  done by Kytos.
- of_flow_manager REST API endpoints and JSON input/output.
- Updated NApps documentation.
- Updated requirements for of_stats.
- of_lldp now ignores lldp packets generated by other applications.
- switch 'lastseen' information is now updated on every new openflow message
  received.

Fixed
=====
- of_topology will no more detect inexistent hosts if the topology has multiple
  switches.
- Several bug fixes.

[0.3.0] - 2017-05-05
********************
Added
=====
- Individual requirements.txt file for NApps.
- Continuous integration with Code Quality Score and test coverage.

Changed
=======
- Installation process of NApps.
- Updated documentation content and templates.
- NApp imports on Kytos now grants easier access to some classes.
- Travis CI replaced by Scrutinizer.
- Logging updated to match changes on Kytos project.
- Updated requirements and dependencies.

Deprecated
==========
- 'author' attribute was renamed to 'username', and will be removed in future
  releases.

Removed
=======
- Installing from source. Now it works only in develop mode. Otherwise,
  kytos-utils shall be used.
- 'long_description' JSON attribute from kytos.json for all NApps. description
  shall be short, and details shall be provided in the README file.

Fixed
=====
- of_core NApp protocol negotiation reviewed to handle correctly Hello, Echo,
  FeaturesRequest and StatsRequest packets.
- Several bug fixes.


[0.2.0] - 2017-03-24
********************
Added
=====
- Python3.6 requirement.
- NApps now can import from other NApps.
- Individual Settings file for NApps.
- Individual documentation for NApps.
- Description field for registered switches.

Changed
=======
- Improved installing process.
- Directory structure of installed/enabled NApps.
- Uptated overall project documentation.
- Updated all Napps metadata.
- of.ipv6disable renamed to of_ipv6drop.
- Updated NApps requirements and dependencies.
- Correct handling of Port Status changes.
- of_stats now supports various link/port speeds in the topology, including
  user-defined.

Deprecated
==========
- Kyco (Kytos Controller) is now called Kytos - every reference was updated to
  match the change.
- Core NApps are now called Kytos NApps.

Removed
=======
- of_liveness NApp.

Fixed
=====
- Tests, coverage and style checks corrected.
- Logger names for NApps.
- Event names corrected to match NApp names.
- NApp names corrected, without dashes or dots.
- Includes all flows and interfaces in the topology, even unused ones.
- Many bug fixes.


[0.1.0] - 2016-11-09
********************
Added
=====
- of.web.topology.layout NApp - Application to manage topology layout on the
  web interface.
- Topology endpoint returning a JSON file with nodes and links.
- A REST API for of.stats NApp.
- of.stats NApp - Collects network statistics and stores it for visualization.
- of.flow_manager NApp - Application to register REST endpoints to manage flows.
- of.lldp NApp - Discovers network topology using Link Layer Discovery Protocol.
- of.ipv6disable NApp - Simple application to disable IPv6 traffic on the
  network.
- kytos.json file with NApp metadata for every NApp.
- of.l2lsloop NApp - An L2 learning switch supporting loops (experimental).
- of.l2ls NApp - An L2 learning switch.
- LICENSE file.
- Pep8 compliance.
- TRAVIS-CI tests and coverage.
- Logging support.
- Setup file compatible with Python Virtual Environment.
- Documentation.
- First Core NApps.

Changed
=======
- Updated requirements for NApps.
- NApps now have a method which will loop itself over intervals.
- of.core NApp totally refactored:

  - Now handles Packet-in messages and generates events correclty.
  - Updates interfaces of switches when receiving features-replies.
  - Updates flows using Flow Stats Request messages.

- Improved NApp installing process.
- Participation of NApps in the handshake process.

Fixed
=====
- Logs for NApps: of.liveness, of.core
- Many bug fixes.


tests/test_port.py
"""Test port statistics."""
from unittest import TestCase
from unittest.mock import patch

from flask import Response

from stats_api import PortStatsAPI


class TestPort(TestCase):
    """Test switch ports."""

    def test_non_existent_port(self):
        """Should return 404."""
        api = PortStatsAPI('non-existent-switch', 123)
        with patch('stats_api.request'):
            response = api.get_stats()
            self.assertIsInstance(response, Response,
                                  'Should be a flask.Response object')


main.py
"""Statistics application."""
from kytos.core import KytosNApp, log, rest
from kytos.core.helpers import listen_to
from pyof.v0x01.controller2switch.stats_request import StatsType

from napps.kytos.of_stats import settings
from napps.kytos.of_stats.stats import FlowStats, PortStats
from napps.kytos.of_stats.stats_api import FlowStatsAPI, PortStatsAPI, StatsAPI


class Main(KytosNApp):
    """Main class for statistics application."""

    def setup(self):
        """Initialize all statistics and set their loop interval."""
        self.execute_as_loop(settings.STATS_INTERVAL)

        # Initialize statistics
        msg_out = self.controller.buffers.msg_out
        self._stats = {StatsType.OFPST_PORT.value: PortStats(msg_out),
                       StatsType.OFPST_FLOW.value: FlowStats(msg_out)}

        StatsAPI.controller = self.controller

    def execute(self):
        """Query all switches sequentially and then sleep before repeating."""
        switches = list(self.controller.switches.values())
        for switch in switches:
            if switch.is_connected():
                self._update_stats(switch)

    def shutdown(self):
        """End of the application."""
        log.debug('Shutting down...')

    def _update_stats(self, switch):
        for stats in self._stats.values():
            if switch.connection is not None:
                stats.request(switch.connection)

    @listen_to('kytos/of_core.v0x01.messages.in.ofpt_stats_reply')
    def listen_v0x01(self, event):
        """Detect the message body type."""
        stats_reply = event.content['message']
        stats_type = stats_reply.body_type
        self._listen(event, stats_type)

    @listen_to('kytos/of_core.v0x04.messages.in.ofpt_multipart_reply')
    def listen_v0x04(self, event):
        """Detect the message body type."""
        multipart_reply = event.content['message']
        stats_type = multipart_reply.multipart_type
        self._listen(event, stats_type)

    def _listen(self, event, stats_type):
        """Listen to all stats reply we deal with.

        Note: v0x01 ``body_type`` and v0x04 ``multipart_type`` have the same
        values.  Besides, both ``msg.body`` have the fields/attributes we use.
        Thus, we can treat them the same way and reuse the code.
        """
        msg = event.content['message']
        if stats_type.value in self._stats:
            stats = self._stats[stats_type.value]
            stats_list = msg.body
            stats.listen(event.source.switch, stats_list)
        else:
            log.debug('No listener for %s = %s in %s.', stats_type.name,
                      stats_type.value, list(self._stats.keys()))

    # REST API

    @rest('v1/<dpid>/ports/<int:port>')
    @staticmethod
    def get_port_stats(dpid, port):
        """Return statistics for ``dpid`` and ``port``."""
        return PortStatsAPI.get_port_stats(dpid, port)

    @rest('v1/<dpid>/ports')
    @staticmethod
    def get_ports_list(dpid):
        """Return ports of ``dpid``."""
        return PortStatsAPI.get_ports_list(dpid)

    @rest('v1/<dpid>/flows/<flow_hash>')
    @staticmethod
    def get_flow_stats(dpid, flow_hash):
        """Return statistics of a flow in ``dpid``."""
        return FlowStatsAPI.get_flow_stats(dpid, flow_hash)

    @rest('v1/<dpid>/flows')
    @staticmethod
    def get_flow_list(dpid):
        """Return all flows of ``dpid``."""
        return FlowStatsAPI.get_flow_list(dpid)

    @rest('v1/<dpid>/ports/<int:port>/random')
    @staticmethod
    def get_random_interface_stats(dpid, port):
        """Fake data for testing."""
        # Ignore dpid and port
        # pylint: disable=unused-argument
        return PortStatsAPI.get_random_port_stats()


settings.py
"""Settings file for the NApp kytos/of_stats."""
from pathlib import Path
from threading import Lock

#: Seconds to wait before asking for more statistics.
#: Delete RRDs everytime this interval is changed
STATS_INTERVAL = 60
# STATS_INTERVAL = 1  # 1 second for testing - check RRD._get_archives()

#: Avoid segmentation fault
rrd_lock = Lock()

# RRD Tool Settings

DIR = Path(__file__).resolve().parent / 'rrd'

# If no new data is supplied for more than *_TIMEOUT* seconds,
# the temperature becomes *UNKNOWN*.
TIMEOUT = 2 * STATS_INTERVAL

# Minimum accepted value
MIN = 0

# Maximum accepted value is the maximum PortStats attribute value.
MAX = 2 ** 64 - 1

# The xfiles factor defines what part of a consolidation interval may be
# made up from *UNKNOWN* data while the consolidated value is still
# regarded as known. It is given as the ratio of allowed *UNKNOWN* PDPs
# to the number of PDPs in the interval. Thus, it ranges from 0 to 1
# (exclusive).
XFF = '0.5'

# How long to keep the data. Accepts s (seconds), m (minutes), h (hours),
# d (days), w (weeks), M (months), and y (years).
# Must be a multiple of consolidation steps.
PERIOD = '30d'


user_speed.py
"""Manage link speeds when OF spec is not enough."""
import json
from os.path import dirname
from pathlib import Path


class UserSpeed:
    """User-defined interface speeds.

    In case there is no matching speed in OF spec or the speed is not correctly
    detected.
    """

    _FILE = Path(dirname(__file__)) / 'user_speed.json'

    def __init__(self):
        """Load user-created file."""
        if self._FILE.exists():
            with self._FILE.open() as user_file:
                # print('UserSpeed', user_file.read())
                self._speed = json.load(user_file)
        else:
            self._speed = {}

    def get_speed(self, dpid, port=None):
        """Return speed in bits/sec or None if not defined by the user.

        Args:
            dpid (str): Switch dpid.
            port (int or str): Port number.
        """
        if not isinstance(port, str):
            port = str(port)
        speed = None
        switch = self._speed.get(dpid)
        if switch is None:
            speed = self._speed.get('default')
        else:
            if port is None or port not in switch:
                speed = switch.get('default')
            else:
                speed = switch.get(port)
        return speed


stats_api.py
"""Module with Classes to handle statistics api."""
import json
from abc import ABCMeta, abstractmethod
from random import randint

from flask import Response, request
from kytos.core import log

from napps.kytos.of_stats.stats import FlowStats, PortStats
from napps.kytos.of_stats.user_speed import UserSpeed


class StatsAPI(metaclass=ABCMeta):
    """Class to answer REST API requests."""

    _rrd = None
    controller = None

    def __init__(self):
        """Initialize instance attributes."""
        self._stats = {}

    def get_points(self, index, n_points=30):
        """Return Flask response for port stats."""
        try:
            data = self._get_points_data(index, n_points)
        except FileNotFoundError as e:
            data = self._get_rrd_not_found_error(e)
        return self._get_response(data)

    def _get_points_data(self, index, n_points):
        start = request.args.get('start')
        end = request.args.get('end')
        return self._fetch(index, start, end, n_points)

    def get_latest(self, fn_items):
        """Return latest stats for items obtained in a switch."""
        switch = self._get_switch()
        if switch is None:
            data = []
        else:
            items = fn_items(switch)
            data = list(self._get_latest_stats(items))
        return self._get_response({'data': data})

    @abstractmethod
    def _get_latest_stats(self, items):
        pass

    def _fetch(self, index, start, end, n_points):
        tstamps, cols, rows = self._rrd.fetch(index, start, end, n_points)
        self._stats = {col: [] for col in cols}
        self._stats['timestamps'] = list(tstamps)
        for row in rows:
            for col, value in zip(cols, row):
                self._stats[col].append(value)
        self._remove_null()
        return {'data': self._stats}

    def _get_switch(self):
        switch = self.controller.get_switch_by_dpid(self._dpid)
        if switch is None:
            log.warning('Switch %s not found in controller', self._dpid[-3:])
        return switch

    def _remove_null(self):
        """Remove a row if all its values are null."""
        nullable_cols = list(self._stats.keys())
        nullable_cols.remove('timestamps')
        n_elements = len(self._stats['timestamps'])

        # Check elements backwards for safe removal
        for i in range(n_elements - 1, -1, -1):
            for col in nullable_cols:
                # Stop if a non-null element is found in the row.
                if self._stats[col][i] is not None:
                    break
            if self._stats[col][i] is not None:
                # Keep current row and check the next one.
                continue
            # Remove the current row from every list
            for lst in self._stats.values():
                lst.pop(i)

    @staticmethod
    def _get_response(dct):
        json_ = json.dumps(dct, sort_keys=True, indent=4)
        # It should be application/vnd.api+json because it follows
        # http://jsonapi.org/format/. However, Firefox doesn't display it and
        # show a download window.
        return Response(json_, mimetype='application/json')

    @staticmethod
    def _get_rrd_not_found_error(exception):
        return {'errors': {
            'status': '404',
            'title': 'Database not found.',
            'detail': str(exception)}}


class PortStatsAPI(StatsAPI):
    """REST API for port statistics."""

    #: key is RRD column, value is a new column name for utilization
    #: percentage.
    _util_cols = {'rx_bytes': 'rx_util',
                  'tx_bytes': 'tx_util'}
    _rrd = PortStats.rrd

    def __init__(self, dpid, port=None):
        """Set dpid and port."""
        super().__init__()
        self._dpid = dpid
        self._port = port

    @classmethod
    def get_port_stats(cls, dpid, port):
        """Get up to 30 points of all statistics of PortStats.

        Includes start and end that are both optional and and must be submitted
        in the form "?start=x&end=y".

        Args:
            dpid (str): Switch dpid.
            port (str, int): Switch port number.
            start (int): Unix timestamp in seconds for the first stats.
                Defaults to the start parameter of the RRD creation.
            end (int): Unix timestamp in seconds for the last stats. Defaults
                to now.
            n_points (int): Return n_points. May return more if there is no
                matching resolution in the RRD file. Defaults to as many points
                as possible.
        """
        api = cls(dpid, port)
        return api.get_stats()

    @classmethod
    def get_ports_list(cls, dpid):
        """List all ports that have statistics and their latest stats.

        Include link utilization.

        Args:
            dpid (str): Switch dpid.
        """
        api = cls(dpid)
        return api.get_list()

    @staticmethod
    def get_random_port_stats():
        stats = {'data': {
            'timestamps': list(range(1508532494, 1508533094, 10)),
            'rx_bytes': [randint(100_000, 1_000_000) for _ in range(30)],
            'tx_bytes': [randint(100_000, 1_000_000) for _ in range(30)],
        }}
        return StatsAPI._get_response(stats)

    def get_list(self):
        """See :meth:`get_ports_list`."""
        return super().get_latest(lambda sw: (sw.interfaces[k]
                                              for k in sorted(sw.interfaces)))

    def _get_latest_stats(self, ifaces):
        for iface in ifaces:
            self._port = iface.port_number
            index = (self._dpid, self._port)
            row = self._rrd.fetch_latest(index)
            row['port'] = self._port
            row['name'] = iface.name
            row['mac'] = iface.address
            row['speed'] = self._get_speed(iface)
            yield self._add_utilization(row, iface)

    def get_stats(self):
        """See :meth:`get_port_stats`."""
        index = (self._dpid, self._port)
        return self.get_points(index)

    def _get_points_data(self, index, n_points):
        """Add port speed to rrd data."""
        response = super()._get_points_data(index, n_points)
        switch = self._get_switch()
        iface = switch.get_interface_by_port_no(self._port)
        response['data']['speed'] = self._get_speed(iface)
        return response

    def _get_speed(self, iface):
        """Update and return interface speed.

        Update controller's interface speed by reading user_speed.json file.
        """
        user = UserSpeed()
        user_speed = user.get_speed(self._dpid, iface.port_number)
        if user_speed != iface.get_custom_speed():
            iface.set_custom_speed(user_speed)
        return iface.speed

    def _add_utilization(self, row, iface):
        """Calculate utilization and also add port number."""
        speed = row['speed']
        if speed is None:
            for util_col in self._util_cols.values():
                row[util_col] = None
            # Shorten dpid for better log readability
            if len(self._dpid) >= 22:
                dpid = self._dpid[:3] + '...' + self._dpid[-3:]
            else:
                dpid = self._dpid
            log.warning('No speed, port %s, dpid %s', self._port, dpid)
        else:
            for bytes_col, util_col in self._util_cols.items():
                row[util_col] = row[bytes_col] / speed  # bytes/sec
        return row


class FlowStatsAPI(StatsAPI):
    """REST API for flow statistics."""

    _rrd = FlowStats.rrd

    def __init__(self, dpid, flow=None):
        """Set dpid and port."""
        super().__init__()
        self._dpid = dpid
        self._flow = flow

    @classmethod
    def get_flow_list(cls, dpid):
        """List all flows that have statistics and their latest stats.

        Args:
            dpid (str): Switch dpid.
        """
        api = cls(dpid)
        return api.get_list()

    @classmethod
    def get_flow_stats(cls, dpid, flow_hash):
        """Return flow statics by its hash.

        Includes start and end that are both optional and and must be submitted
        in the form "?start=x&end=y".

        Args:
            dpid (str): Switch dpid.
            flow_hash (str): Flow hash.
            start (int): Unix timestamp in seconds for the first stats.
                Defaults to the start parameter of the RRD creation.
            end (int): Unix timestamp in seconds for the last stats. Defaults
                to now.
        """
        api = cls(dpid, flow_hash)
        return api.get_stats()

    def get_list(self):
        """See :meth:`get_flow_list`."""
        return super().get_latest(lambda sw: sorted(sw.flows,
                                                    key=lambda f: f.id))

    def _get_latest_stats(self, flows):
        for flow in flows:
            index = (self._dpid, flow.id)
            rrd_data = self._rrd.fetch_latest(index)
            stats = {}
            stats['Bps'] = rrd_data.get('byte_count', 0)
            stats['pps'] = rrd_data.get('packet_count', 0)
            dct = flow.as_dict()
            # Make it JS friendly
            dct['id'] = dct.pop('id')
            dct['stats'] = stats
            yield dct

    def get_stats(self):
        """See :meth:`get_flow_stats`."""
        index = (self._dpid, self._flow)
        return super().get_points(index)
